presto-main/src/main/java/com/facebook/presto/type/DecimalOperators.java

```diff
@@ -336,146 +340,94 @@ private static SqlScalarFunction decimalDivideOperator()
     private static int divideRescaleFactor(SqlScalarFunctionBuilder.SpecializeContext context)
     {
         DecimalType returnType = (DecimalType) context.getReturnType();
-        // +1 because we want to do computations with one extra decimal field to be able to handle rounding of the result.
-        return (int) (returnType.getScale() - context.getLiteral("a_scale") + context.getLiteral("b_scale") + 1);
+        return Ints.checkedCast(returnType.getScale() - context.getLiteral("a_scale") + context.getLiteral("b_scale"));
     }
 
     @UsedByGeneratedCode
-    public static long divideShortShortShort(long a, long b, long aRescale)
+    public static long divideShortShortShort(long dividend, long divisor, long scale)
     {
-        try {
-            long result = a * aRescale / b;
-            if (result > 0) {
-                if (result % 10 >= 5) {
-                    return result / 10 + 1;
-                }
-                else {
-                    return result / 10;
-                }
-            }
-            else {
-                if (result % 10 <= -5) {
-                    return result / 10 - 1;
-                }
-                else {
-                    return result / 10;
-                }
-            }
+        if (divisor == 0) {
+            throw new PrestoException(DIVISION_BY_ZERO, "Division by zero");
         }
-        catch (ArithmeticException e) {
-            throw new PrestoException(DIVISION_BY_ZERO, e);
+
+        int resultSignum = signum(dividend) * signum(divisor);
+        if (resultSignum == 0) {
+            return 0;
         }
-    }
 
-    @UsedByGeneratedCode
-    public static long divideShortLongShort(long a, Slice b, BigInteger aRescale)
-    {
-        BigInteger aBigInteger = BigInteger.valueOf(a);
-        BigInteger bBigInteger = Decimals.decodeUnscaledValue(b);
-        return internalDivideShortResult(aBigInteger, bBigInteger, aRescale);
+        long unsignedDividend = abs(dividend);
+        long unsignedDivisor = abs(divisor);
+
+        long rescaledUnsignedDividend = unsignedDividend * scale;
+        long quotient = rescaledUnsignedDividend / unsignedDivisor;
+        long remainder = rescaledUnsignedDividend - (quotient * unsignedDivisor);
+
+        if (Long.compareUnsigned(remainder * 2, unsignedDivisor) >= 0) {
```

Comment:
Sorry, I wasn't asking about `compareUnsigned` specifically. I was wondering why we even need to do this comparison (remainder * 2 vs divisor) and adjust quotient.